src/backend/marsha/core/tests/test_api_liveregistration.py

```diff
@@ -0,0 +1,779 @@
+        starting_at = timezone.now() + timedelta(days=5)
+        video = VideoFactory(starting_at=starting_at)
+        self.assertTrue(video.is_scheduled)
+        jwt_token = AccessToken()
+        jwt_token.payload["resource_id"] = str(video.id)
+
+        response = self.client.post(
+            "/api/liveregistrations/",
+            {"should_send_reminders": True},
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {jwt_token}",
+        )
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(
+            json.loads(response.content),
+            {"email": ["This field is required."]},
+        )
+
+    def test_api_liveregistration_create_cant_register_two_identical_emails(self):
+        """Same email can't be used for the same video."""
+        starting_at = timezone.now() + timedelta(days=5)
+        video = VideoFactory(starting_at=starting_at)
+        LiveRegistrationFactory(email="salome@test-fun-mooc.fr", video=video)
+        self.assertTrue(video.is_scheduled)
+        jwt_token = AccessToken()
+        jwt_token.payload["resource_id"] = str(video.id)
+
+        response = self.client.post(
+            "/api/liveregistrations/",
+            {"email": "salome@test-fun-mooc.fr", "should_send_reminders": True},
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {jwt_token}",
+        )
+        self.assertEqual(response.status_code, 400)
+
+    def test_api_liveregistration_create_cant_register_same_lti_info_different_email(
+        self,
+    ):
+        """Same combinaison video and lti can't be used for different registrations."""
+        starting_at = timezone.now() + timedelta(days=5)
+        video = VideoFactory(starting_at=starting_at)
+        LiveRegistrationFactory(
+            email="salome@test-fun-mooc.fr",
+            video=video,
+            lti_user_id="56255f3807599c377bf0e5bf072359fd",
+            lti_consumer_site=video.playlist.consumer_site,
+        )
+        self.assertTrue(video.is_scheduled)
+        jwt_token = AccessToken()
+        jwt_token.payload["resource_id"] = str(video.id)
+        jwt_token.payload["context_id"] = str(video.playlist.lti_id)
+        jwt_token.payload["user"] = {
+            "email": None,
+            "id": "56255f3807599c377bf0e5bf072359fd",
+            "username": "Token",
+        }
+        response = self.client.post(
+            "/api/liveregistrations/",
+            {"email": "balou@test-fun-mooc.fr", "should_send_reminders": True},
+            content_type="application/json",
+            HTTP_AUTHORIZATION=f"Bearer {jwt_token}",
+        )
+        self.assertEqual(
+            json.loads(response.content),
+            {"error": "IntegrityError"},
+        )
+        self.assertEqual(response.status_code, 400)
+
+    def test_api_liveregistration_create_can_register_two_identical_emails_different_video(
```

Comment:
This test is randomly failing with this error

```
E       AssertionError: {'ema[36 chars]': '42f4d958-3f8b-414c-bef7-31f18e8a4050', 'lt[119 chars]5c7'} != {'ema[36 chars]': '4651f6ef-c65a-4543-9fc7-a8495df439f3', 'lt[119 chars]5c7'}
E         {'email': 'chantal@test-fun-mooc.fr',
E       -  'id': '42f4d958-3f8b-414c-bef7-31f18e8a4050',
E       +  'id': '4651f6ef-c65a-4543-9fc7-a8495df439f3',
E          'lti_consumer_site': None,
E          'lti_user_id': None,
E          'should_send_reminders': True,
E          'video': '6f7f76a0-17b4-484a-997f-09afcda2a5c7'}

marsha/core/tests/test_api_liveregistration.py:702: AssertionError
```

Reply:
yes wdyt about it, because it makes no sense to me, randomly the token doesn't seem to be updated, which would cause a problem if it really happens, I think it's good to take a serious look at it. I've left the test on purpose, on my side I didn't find an explanation.

Reply:
```
        saved_liveregistration = LiveRegistrationFactory(email="chantal@test-fun-mooc.fr", video=video)
        jwt_token2 = AccessToken()
        jwt_token2.payload["resource_id"] = str(video2.id)

        # With the same email but other video, liveregistration is possible
        response = self.client.post(
            "/api/liveregistrations/",
            {"email": "chantal@test-fun-mooc.fr", "should_send_reminders": True},
            HTTP_AUTHORIZATION=f"Bearer {jwt_token2}",
            content_type="application/json",
        )
        self.assertEqual(response.status_code, 201)
        liveregistration = LiveRegistration.objects.last()
        print(saved_liveregistration)
        print(liveregistration)
```

`LiveRegistration.objects.last()` returns the previously saved live registration. Not the one created in the request.

Reply:
LiveRegistration.objects.last() should return the last object created, and the request creates an object so it should return this object, the fact that it doesn't doesn't make sens to me.

Reply:
Actually, .last() returns the last object as per the ordering of the query. So in your case the default ordering... which is by the primary key. But in Marsha, the primary key is a UUIDField, so the sorting is completely random: https://github.com/openfun/marsha/blob/master/src/backend/marsha/core/models/base.py#L76